instrumentation/servlet/src/main/java/brave/servlet/TracingFilter.java

```diff
@@ -0,0 +1,88 @@
+package brave.servlet;
+
+import brave.Span;
+import brave.Tracer;
+import brave.Tracing;
+import brave.http.HttpServerParser;
+import brave.propagation.TraceContext;
+import java.io.IOException;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public final class TracingFilter implements Filter {
+  /** Creates a tracing filter with defaults. Use {@link #newBuilder(Tracing)} to customize. */
+  public static TracingFilter create(Tracing tracing) {
+    return newBuilder(tracing).build();
+  }
+
+  public static Builder newBuilder(Tracing tracing) {
+    return new Builder(tracing);
+  }
+
+  public static final class Builder {
+    final Tracing tracing;
+    HttpServerParser<HttpServletRequest, HttpServletResponse> parser = new HttpServerParser<>();
+
+    Builder(Tracing tracing) { // intentionally hidden
+      if (tracing == null) throw new NullPointerException("tracing == null");
+      this.tracing = tracing;
+    }
+
+    public Builder parser(HttpServerParser<HttpServletRequest, HttpServletResponse> parser) {
+      if (parser == null) throw new NullPointerException("parser == null");
+      this.parser = parser;
+      return this;
+    }
+
+    public TracingFilter build() {
+      return new TracingFilter(this);
+    }
+  }
+
+  final ServletRuntime runtime = ServletRuntime.get();
+  final Tracer tracer;
+  final HttpServletHandler handler;
+  final TraceContext.Extractor<HttpServletRequest> extractor;
+
+  TracingFilter(Builder builder) {
+    tracer = builder.tracing.tracer();
+    handler = new HttpServletHandler(builder.parser);
+    extractor = builder.tracing.propagation().extractor(HttpServletRequest::getHeader);
+  }
+
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain)
```

Comment:
@devinsba this servlet filter passes all tests sync and async. Will add an invoker test to make sure it also works on servlet 2.5 in practice (we have had people still using that, yes).